### PR TITLE
provide default value for --files when none is provided

### LINKED ISSUE #169: rebar3: Crash upon `rebar3 fmt **`
I'm sure this syntax is wrong, but it was the first thing I tried when I ran `rebar3 fmt` and was told about `no files provided to format`.

**What was I expecting?**

As per [filelib:wildcard/1](https://erlang.org/doc/man/filelib.html#wildcard-1), to "match all files and zero or more directories and subdirectories."

Using `master`.

By the way:

* `/**`
* `src/**`

also crash.

**What was I expecting?**

That the formatter would execute, and if files had to be skipped, they would be - with a warning, stating that it can't format so and so, because "unknown file type" -, instead of crashing 😄 

Oh, I now see that it's not expected to crash, since it's in the `README.md`. Anyway, here's my report...

### PR DESCRIPTION
**Pull request by awalterschulze**
Fixes #169

### CODE REVIEW

File: src/erlfmt_cli.erl [hunk 1]
```diff
@@ -58,7 +61,23 @@ do(Name, Opts) ->
 do(Name, PreferOpts, DefaultOpts) ->
     PreferParsed = parse_opts(PreferOpts),
     DefaultParsed = parse_opts(DefaultOpts),
-    Parsed = resolve_parsed(PreferParsed, DefaultParsed),
+    case {specified_files(PreferOpts), PreferParsed} of
+        {SpecifiedFiles, {format, [], _}} when SpecifiedFiles =/= [] ->
```
Comment (michalmuskala) [inline src/erlfmt_cli.erl:65]: If `PreferParsed` returns an empty list of files to format, do we need to run the `specified_files` check? What's the case when they would return different values?

**Comment by michalmuskala** [comment]
What's the behaviour when you just run `erlfmt` without any options? Today it prints the help, will it now format things to stdout? I think it would be worth it to preserve the current behaviour. We could provide the default files when one of the "action" flags is passed to erlfmt, like `-w`, `-c`, or `-o`

File: src/erlfmt_cli.erl (hunk 1 above)
Comment (awalterschulze) [reply src/erlfmt_cli.erl:65]: If you do not specify files, we are going to add the default files, but we do not want to add these default files in cases where you specified files that just didn't match any existing files.

**Comment by awalterschulze** [comment]
Yes it prints to stdout.  I think writing files to stdout is harmless and we do provide classic `-h` and `--help` flags for help.
We could enable this only for those specific flags, but then there is no way to invoke the stdout behaviour for the default files.

**Comment by michalmuskala** [comment]
I would argue stdout with multiple files is quite useless in the first place - you can't tell where one file ends, the other starts, similar to `cat`ing multiple files together.

**Comment by awalterschulze** [comment]
Cool sounds good, I will update it to only add default files, given at least one other parameter.

**Comment by paulo-ferraz-oliveira** [comment]
Would PR title "provide default value for --files when none is provided" be clearer? (by reading it, I'm under the impression actual files are going to be added) 😄

**Comment by paulo-ferraz-oliveira** [comment]
The README still says `rebar3 fmt` will "format all the files in your project" (it doesn't), but my main issue is that `rebar3 fmt src/*.erl` doesn't work. Isn't this syntax OK? (I'm imagining you use `wildcard` behind the scenes)

**Review by michalmuskala — APPROVED** [review]

**Comment by awalterschulze** [comment]
I have now also tested `rebar3 fmt src/*.erl` and it formats my .erl files in my local example project.

**Comment by awalterschulze** [comment]
I have also checked the Readme and it says "**Now** you can format all the files in your project by running:" which implies that you have setup the rebar config.

**Comment by awalterschulze** [comment]
@ilya-klyuchnikov I think you also wanted this feature.

**Comment by michalmuskala** [comment]
There are two layers of file expansion.

When you call `rebar3 fmt src/*.erl` it's the shell that expands the files. When you call `rebar3 fmt 'src/*.erl'` it's erlfmt using `filelib:wildcard/1` - this is generally needed since windows doesn't do wildcard expansion in the shell and we also want to support some convenient CLI for users there. `filelib:wildcard/1` is also used for config from `rebar.config`.

**Comment by paulo-ferraz-oliveira** [comment]
Thanks for the explanation @michalmuskala. I understand the reasoning now, but having to use `'` is still surprising 😄 (it worked though).

**Comment by awalterschulze** [comment]
@michalmuskala wow I never knew that.  Quite embarrassing. 
I had to check and found that the behaviour is the same in Go.  I think I avoided asking this question, because I usually used Go's special `...` syntax, which doesn't get expanded by unix and then we could expand it with a function call.
I always thought the `...` was weirdly inconsistent, but now I get that, that was on purpose to avoid the unix expansion.

**Pull request merged**